File: frontend/src/app/services/patient.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Patient } from '../models/patient';

@Injectable({
  providedIn: 'root',
})
export class PatientService {
  private http = inject(HttpClient);
  private patientServiceBaseURL = 'http://localhost:8080/patients';

  /**
   * Retrieves all patients.
   */
  public getPatients(): Observable<Patient[]> {
    console.log('patient service base url', this.patientServiceBaseURL);
    return this.http.get<Patient[]>(this.patientServiceBaseURL);
  }

  /**
   * Retrieves a patient by ID.
   * @param id - The ID of the patient to retrieve.
   */
  public getPatient(id: number): Observable<Patient> {
    return this.http.get<Patient>(`${this.patientServiceBaseURL}/${id}`);
  }

  /**
   * Creates a new patient.
   * @param patient - The patient object to create.
   */
  public createPatient(patient: Patient): Observable<Patient> {
    return this.http.post<Patient>(this.patientServiceBaseURL, patient);
  }

  /**
   * Updates an existing patient.
   * @param patient - The patient object to update.
   */
  public updatePatient(patient: Patient): Observable<Patient> {
    return this.http.put<Patient>(
      `${this.patientServiceBaseURL}/${patient.id}`,
      patient
    );
  }
}

File: patient-service/src/main/java/com/diabetesrisk/patient_service/service/PatientService.java
package com.diabetesrisk.patient_service.service;

import com.diabetesrisk.patient_service.model.Patient;
import com.diabetesrisk.patient_service.model.PatientDto;

import java.util.List;

public interface PatientService {

    /**
     * Get all patients.
     *
     * @return List of patients
     */
    List<Patient> getPatients();

    /**
     * Get a patient by ID.
     *
     * @param id the ID of the patient
     * @return the patient
     */
    Patient getPatient(int id);

    /**
     * Save a patient.
     *
     * @param patient the patient to save
     * @return the saved patient
     */
    Patient savePatient(Patient patient);

    /**
     * Update a patient.
     *
     * @param patient the patient to update
     * @return the updated patient
     */
    Patient updatePatient(Patient patient);

    /**
     * Get a patient DTO by ID.
     *
     * @param id the ID of the patient
     * @return the patient DTO
     */
    PatientDto getPatientDto(Integer id);
}
